File: src/main/java/com/arcao/geocaching4locus/settings/fragment/FilterPreferenceFragment.java
package com.arcao.geocaching4locus.settings.fragment;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.EditTextPreference;
import android.preference.Preference;
import android.text.method.DigitsKeyListener;
import android.widget.EditText;
import com.arcao.geocaching.api.data.type.ContainerType;
import com.arcao.geocaching.api.data.type.GeocacheType;
import com.arcao.geocaching4locus.App;
import com.arcao.geocaching4locus.R;
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment;
import com.arcao.geocaching4locus.base.constants.PrefConstants;

public class FilterPreferenceFragment extends AbstractPreferenceFragment {
	public static final String TEXT_VALUE_SEPARATOR = ", ";

	private boolean mPremiumMember;
	private boolean mImperialUnits;

	@Override
	public void onCreate(Bundle paramBundle) {
		super.onCreate(paramBundle);

		// Load the preferences from an XML resource
		addPreferencesFromResource(R.xml.preference_category_filter);

		mPremiumMember = App.get(getActivity()).getAccountManager().isPremium();
		mImperialUnits = mPrefs.getBoolean(PrefConstants.IMPERIAL_UNITS, false);
	}

	@Override
	protected void preparePreference() {
		prepareCacheTypePreference();
		prepareContainerTypePreference();
		prepareDifficultyPreference();
		prepareTerrainPreference();
		prepareDistancePreference();
	}

	@Override
	public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
		super.onSharedPreferenceChanged(sharedPreferences, key);

		if (key == null)
			return;

		switch (key) {
			case FILTER_DISTANCE:
				final EditTextPreference distancePreference = findPreference(key, EditTextPreference.class);
				if (mImperialUnits) {
					distancePreference.setSummary(preparePreferenceSummary(distancePreference.getText() + UNIT_MILES, R.string.pref_distance_summary_miles));
				} else {
					distancePreference.setSummary(preparePreferenceSummary(distancePreference.getText() + UNIT_KM, R.string.pref_distance_summary_km));
				}
				break;
		}
	}

	private void prepareCacheTypePreference() {
		final Preference cacheTypeFilterScreen = findPreference(FILTER_CACHE_TYPE, Preference.class);
		cacheTypeFilterScreen.setEnabled(mPremiumMember);
		cacheTypeFilterScreen.setSummary(mPremiumMember ? prepareCacheTypeSummary() : prepareCacheTypeSummaryBasicMember());
		if (!mPremiumMember) {
			applyPremiumTitleSign(cacheTypeFilterScreen);
		}
	}

	private void prepareContainerTypePreference() {
		final Preference containerTypeFilterScreen = findPreference(FILTER_CONTAINER_TYPE, Preference.class);
		containerTypeFilterScreen.setEnabled(mPremiumMember);
		containerTypeFilterScreen.setSummary(
				mPremiumMember ? prepareContainerTypeSummary() : prepareContainerTypeSummaryBasicMember());
		if (!mPremiumMember) {
			applyPremiumTitleSign(containerTypeFilterScreen);
		}
	}

	private void prepareDifficultyPreference() {
		final Preference difficultyPreference = findPreference(FILTER_DIFFICULTY, Preference.class);
		difficultyPreference.setEnabled(mPremiumMember);

		String difficultyMin = "1";
		String difficultyMax = "5";

		if (mPremiumMember) {
			difficultyMin = mPrefs.getString(PrefConstants.FILTER_DIFFICULTY_MIN, difficultyMin);
			difficultyMax = mPrefs.getString(PrefConstants.FILTER_DIFFICULTY_MAX, difficultyMax);
		} else {
			applyPremiumTitleSign(difficultyPreference);
		}

		difficultyPreference.setSummary(prepareRatingSummary(difficultyMin, difficultyMax));
	}

	private void prepareTerrainPreference() {
		final Preference terrainPreference = findPreference(FILTER_TERRAIN, Preference.class);
		terrainPreference.setEnabled(mPremiumMember);

		String terrainMin = "1";
		String terrainMax = "5";

		if (mPremiumMember) {
			terrainMin = mPrefs.getString(PrefConstants.FILTER_TERRAIN_MIN, terrainMin);
			terrainMax = mPrefs.getString(PrefConstants.FILTER_TERRAIN_MAX, terrainMax);
		} else {
			applyPremiumTitleSign(terrainPreference);
		}

		terrainPreference.setSummary(prepareRatingSummary(terrainMin, terrainMax));
	}

	private void prepareDistancePreference() {
		final EditTextPreference distancePreference = findPreference(FILTER_DISTANCE, EditTextPreference.class);
		final EditText filterDistanceEditText = distancePreference.getEditText();
		filterDistanceEditText.setKeyListener(DigitsKeyListener.getInstance(false, true));

		// set summary text
		if (!mImperialUnits) {
			distancePreference.setSummary(preparePreferenceSummary(distancePreference.getText() + UNIT_KM, R.string.pref_distance_summary_km));
		} else {
			distancePreference.setDialogMessage(R.string.pref_distance_summary_miles);
			distancePreference.setSummary(preparePreferenceSummary(distancePreference.getText() + UNIT_MILES, R.string.pref_distance_summary_miles));
		}
	}

	private CharSequence prepareRatingSummary(CharSequence min, CharSequence max) {
		return preparePreferenceSummary(min + " - " + max, 0);
	}

	private CharSequence prepareCacheTypeSummary() {
		StringBuilder sb = new StringBuilder();

		boolean allChecked = true;
		boolean noneChecked = true;

		final int len = GeocacheType.values().length;
		for (int i = 0; i < len; i++) {
			if (mPrefs.getBoolean(PrefConstants.FILTER_CACHE_TYPE_PREFIX + i, true)) {
				noneChecked = false;
			} else {
				allChecked = false;
			}
		}

		if (allChecked || noneChecked) {
			sb.append(getString(R.string.pref_geocache_type_all));
		} else {
			for (int i = 0; i < len; i++) {
				if (mPrefs.getBoolean(PrefConstants.FILTER_CACHE_TYPE_PREFIX + i, true)) {
					if (sb.length() != 0) sb.append(TEXT_VALUE_SEPARATOR);
					sb.append(shortCacheTypeName[i]);
				}
			}
		}

		return preparePreferenceSummary(sb.toString(), 0);
	}

	private CharSequence prepareCacheTypeSummaryBasicMember() {
		return preparePreferenceSummary(shortCacheTypeName[GeocacheType.Traditional.ordinal()] +
				TEXT_VALUE_SEPARATOR +
				shortCacheTypeName[GeocacheType.Event.ordinal()] +
				TEXT_VALUE_SEPARATOR +
				shortCacheTypeName[GeocacheType.CacheInTrashOutEvent.ordinal()],
				0);
	}

	private CharSequence prepareContainerTypeSummary() {
		StringBuilder sb = new StringBuilder();

		final int len = ContainerType.values().length;
		for (int i = 0; i < len; i++) {
			if (mPrefs.getBoolean(PrefConstants.FILTER_CONTAINER_TYPE_PREFIX + i, true)) {
				if (sb.length() != 0) sb.append(TEXT_VALUE_SEPARATOR);
				sb.append(shortContainerTypeName[i]);
			}
		}

		if (sb.length() == 0) {
			for (int i = 0; i < len; i++) {
				if (sb.length() != 0) sb.append(TEXT_VALUE_SEPARATOR);
				sb.append(shortContainerTypeName[i]);
			}
		}

		return preparePreferenceSummary(sb.toString(), 0);
	}

	private CharSequence prepareContainerTypeSummaryBasicMember() {
		StringBuilder sb = new StringBuilder();

		final int len = ContainerType.values().length;
		for (int i = 0; i < len; i++) {
			if (sb.length() != 0) sb.append(", ");
			sb.append(shortContainerTypeName[i]);
		}

		return preparePreferenceSummary(sb.toString(), 0);
	}

}

File: app/src/main/java/com/arcao/geocaching4locus/settings/fragment/FilterPreferenceFragment.kt
package com.arcao.geocaching4locus.settings.fragment

import android.content.SharedPreferences
import androidx.preference.EditTextPreference
import androidx.preference.Preference
import com.arcao.geocaching4locus.R
import com.arcao.geocaching4locus.authentication.util.isPremium
import com.arcao.geocaching4locus.base.constants.AppConstants
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_CACHE_TYPE
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_CACHE_TYPE_PREFIX
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_CONTAINER_TYPE
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_CONTAINER_TYPE_PREFIX
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_DIFFICULTY
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_DIFFICULTY_MAX
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_DIFFICULTY_MIN
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_DISTANCE
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_TERRAIN
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_TERRAIN_MAX
import com.arcao.geocaching4locus.base.constants.PrefConstants.FILTER_TERRAIN_MIN
import com.arcao.geocaching4locus.base.constants.PrefConstants.IMPERIAL_UNITS
import com.arcao.geocaching4locus.base.constants.PrefConstants.SHORT_CACHE_TYPE_NAMES
import com.arcao.geocaching4locus.base.constants.PrefConstants.SHORT_CONTAINER_TYPE_NAMES
import com.arcao.geocaching4locus.base.constants.PrefConstants.UNIT_KM
import com.arcao.geocaching4locus.base.constants.PrefConstants.UNIT_MILES
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment
import com.arcao.geocaching4locus.data.account.AccountManager
import org.koin.android.ext.android.get

class FilterPreferenceFragment : AbstractPreferenceFragment() {
    private val premiumMember: Boolean by lazy {
        get<AccountManager>().isPremium
    }

    private val imperialUnits: Boolean by lazy {
        preferences.getBoolean(IMPERIAL_UNITS, false)
    }

    override val preferenceResource: Int
        get() = R.xml.preference_category_filter

    override fun preparePreference() {
        super.preparePreference()

        prepareCacheTypePreference()
        prepareContainerTypePreference()
        prepareDifficultyPreference()
        prepareTerrainPreference()
        prepareDistancePreference()
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        super.onSharedPreferenceChanged(sharedPreferences, key)

        when (key) {
            FILTER_DISTANCE -> {
                preference<EditTextPreference>(key).apply {
                    summary = if (imperialUnits) {
                        preparePreferenceSummary(text + UNIT_MILES, R.string.pref_distance_summary_miles)
                    } else {
                        preparePreferenceSummary(text + UNIT_KM, R.string.pref_distance_summary_km)
                    }
                }
            }
        }
    }

    private fun prepareCacheTypePreference() {
        preference<Preference>(FILTER_CACHE_TYPE).apply {
            isEnabled = premiumMember
            summary = if (premiumMember) prepareCacheTypeSummary() else prepareCacheTypeSummaryBasicMember()
            if (!premiumMember) {
                applyPremiumTitleSign(this)
            }
        }
    }

    private fun prepareContainerTypePreference() {
        preference<Preference>(FILTER_CONTAINER_TYPE).apply {
            isEnabled = premiumMember
            summary = if (premiumMember) prepareContainerTypeSummary() else prepareContainerTypeSummaryBasicMember()
            if (!premiumMember) {
                applyPremiumTitleSign(this)
            }
        }
    }

    private fun prepareDifficultyPreference() {
        preference<Preference>(FILTER_DIFFICULTY).apply {
            isEnabled = premiumMember

            var difficultyMin = "1"
            var difficultyMax = "5"

            if (premiumMember) {
                difficultyMin = preferences.getString(FILTER_DIFFICULTY_MIN, difficultyMin)!!
                difficultyMax = preferences.getString(FILTER_DIFFICULTY_MAX, difficultyMax)!!
            } else {
                applyPremiumTitleSign(this)
            }

            summary = prepareRatingSummary(difficultyMin, difficultyMax)
        }
    }

    private fun prepareTerrainPreference() {
        preference<Preference>(FILTER_TERRAIN).apply {
            isEnabled = premiumMember

            var terrainMin = "1"
            var terrainMax = "5"

            if (premiumMember) {
                terrainMin = preferences.getString(FILTER_TERRAIN_MIN, terrainMin)!!
                terrainMax = preferences.getString(FILTER_TERRAIN_MAX, terrainMax)!!
            } else {
                applyPremiumTitleSign(this)
            }

            summary = prepareRatingSummary(terrainMin, terrainMax)
        }
    }

    private fun prepareDistancePreference() {
        preference<EditTextPreference>(FILTER_DISTANCE).apply {
            // set summary text
            summary = if (!imperialUnits) {
                preparePreferenceSummary(text + UNIT_KM, R.string.pref_distance_summary_km)
            } else {
                setDialogMessage(R.string.pref_distance_summary_miles)
                preparePreferenceSummary(text + UNIT_MILES, R.string.pref_distance_summary_miles)
            }
        }
    }

    private fun prepareRatingSummary(min: CharSequence, max: CharSequence): CharSequence {
        return preparePreferenceSummary("$min - $max", 0)
    }

    private fun prepareCacheTypeSummary(): CharSequence {
        val sb = StringBuilder()

        var allChecked = true
        var noneChecked = true

        val len = AppConstants.GEOCACHE_TYPES.size
        for (i in 0 until len) {
            if (preferences.getBoolean(FILTER_CACHE_TYPE_PREFIX + i, true)) {
                noneChecked = false
            } else {
                allChecked = false
            }
        }

        if (allChecked || noneChecked) {
            sb.append(getString(R.string.pref_geocache_type_all))
        } else {
            for (i in 0 until len) {
                if (preferences.getBoolean(FILTER_CACHE_TYPE_PREFIX + i, true)) {
                    if (sb.isNotEmpty())
                        sb.append(TEXT_VALUE_SEPARATOR)
                    sb.append(SHORT_CACHE_TYPE_NAMES[i])
                }
            }
        }

        return preparePreferenceSummary(sb.toString(), 0)
    }

    private fun prepareCacheTypeSummaryBasicMember(): CharSequence {
        return preparePreferenceSummary(
            SHORT_CACHE_TYPE_NAMES[0] +
                TEXT_VALUE_SEPARATOR +
                SHORT_CACHE_TYPE_NAMES[8] +
                TEXT_VALUE_SEPARATOR +
                SHORT_CACHE_TYPE_NAMES[10],
            0
        )
    }

    private fun prepareContainerTypeSummary(): CharSequence {
        val sb = StringBuilder()

        val len = AppConstants.GEOCACHE_SIZES.size
        for (i in 0 until len) {
            if (preferences.getBoolean(FILTER_CONTAINER_TYPE_PREFIX + i, true)) {
                if (sb.isNotEmpty())
                    sb.append(TEXT_VALUE_SEPARATOR)
                sb.append(SHORT_CONTAINER_TYPE_NAMES[i])
            }
        }

        if (sb.isEmpty()) {
            for (i in 0 until len) {
                if (sb.isNotEmpty())
                    sb.append(TEXT_VALUE_SEPARATOR)
                sb.append(SHORT_CONTAINER_TYPE_NAMES[i])
            }
        }

        return preparePreferenceSummary(sb.toString(), 0)
    }

    private fun prepareContainerTypeSummaryBasicMember(): CharSequence {
        val sb = StringBuilder()

        val len = AppConstants.GEOCACHE_SIZES.size
        for (i in 0 until len) {
            if (sb.isNotEmpty())
                sb.append(TEXT_VALUE_SEPARATOR)
            sb.append(SHORT_CONTAINER_TYPE_NAMES[i])
        }

        return preparePreferenceSummary(sb.toString(), 0)
    }

    companion object {
        private const val TEXT_VALUE_SEPARATOR = ", "
    }
}
